### PR TITLE
[ntuple] fixes and more testing for std::variant

### LINKED ISSUE #15534: RNTuple: fields with mixed STL types sometimes fail to be filled
### Check duplicate issues.

- [X] Checked for duplicates

### Description

When having fields with mixed STL types sometimes you run into issues when trying to fill them. I've attached an example below.

### Reproducer

This is a minimal script that shows that issue.

```cpp
#include <ROOT/RNTupleModel.hxx>
#include <ROOT/RNTupleWriter.hxx>
 
#include <TSystem.h>
 
#include <vector>
#include <variant>
#include <optional>
 
using RNTupleModel = ROOT::Experimental::RNTupleModel;
using RNTupleWriter = ROOT::Experimental::RNTupleWriter;
 
void ntpl_issue()
{
   auto model = RNTupleModel::Create();
 
   auto fldVvar = model->MakeField<std::vector<std::variant<std::optional<int>,float>>>("vvar");
 
   auto ntuple = RNTupleWriter::Recreate(std::move(model), "F", "ntpl_issue.root");
 
   for (int i = 0; i < 10; i++) {
      fldVvar->clear();
 
      for (int j = 0; j < 5; ++j) {
         std::variant<std::optional<int>,float> var(1.0);
         fldVvar->emplace_back(var);
      }
 
      ntuple->Fill();
   }
}
```

And this is the error that it produces.
```
Fatal: (typedValue->size() % fItemSize) == 0 violated at line 2432 of `.../root_src/tree/ntuple/v7/src/RField.cxx'
aborting
```

Another way to get it to fail is by using `std::vector<std::variant<std::atomic<int>,float>>`.

### ROOT version

6.33/01 (commit eef2244)

### Installation method

Built from source

### Operating system

macOS

### Additional context

I found this issue while trying to generate `std::variant` types in an invalid state by doing something like this.

```cpp
struct S {
   operator int() { throw 42; }
};

std::variant<int,float> var;
try {
  var = S();
} catch (int) {}
fldVvar->emplace_back(var);
```

The spec indicates that invalid states are supported, as shown in the line below. I wanted to ask what's the reason for them being supported given that they are not really supposed to be "legal", and it seems like one can't successfully generate an RNTuple with an invalid state. It either ends up initializing the first variant or it crashes.

https://github.com/root-project/root/blob/95307116ca3dd811ac1b5e496ad7f9828402dc51/tree/ntuple/v7/doc/specifications.md?plain=1#L780

### PR DESCRIPTION
**Pull request by jblomer**
Fixes #15534

### CODE REVIEW

**Review by silverweed — APPROVED** [review]
LGTM

File: tree/ntuple/v7/src/RField.cxx [hunk 1]
```diff
@@ -3218,7 +3235,18 @@ ROOT::Experimental::RVariantField::RVariantField(std::string_view fieldName,
...
+   // With certain template parameters, the union of members of an std::variant starts at an offset > 0.
+   // For instance, std::variant<std::optional<int>> on macOS.
+   auto cl = TClass::GetClass(GetTypeName().c_str());
+   assert(cl);
+   auto dm = reinterpret_cast<TDataMember *>(cl->GetListOfDataMembers()->First());
+   if (dm)
+      fVariantOffset = dm->GetOffset();
+
+   const auto tagSize = GetVariantTagSize();
+   const auto padding = tagSize - (fMaxItemSize % tagSize);
```
Comment (silverweed) [inline tree/ntuple/v7/src/RField.cxx:3248]: If `tagSize` is always a power of two the modulo operation can be written more efficiently as `fMaxItemSize & (tagSize - 1)`

File: tree/ntuple/v7/src/RField.cxx [hunk 2]
```diff
@@ -3318,20 +3349,20 @@ std::unique_ptr<ROOT::Experimental::RFieldBase::RDeleter> ROOT::Experimental::RV
...
-   const auto actualSize = fMaxItemSize + 1;
-   auto padding = 0;
-   if (alignment > 1) {
-      auto remainder = actualSize % alignment;
-      if (remainder != 0)
-         padding = alignment - remainder;
-   }
-   return actualSize + padding;
+   const auto actualSize = fTagOffset + GetVariantTagSize();
+   const auto padding = alignment - (actualSize % alignment);
```
Comment (silverweed) [inline tree/ntuple/v7/src/RField.cxx:3364]: Same comment about modulo of power of two as above: `(a % b) == a & (b - 1)`

**Review by silverweed — CHANGES_REQUESTED** [review]

File: tree/ntuple/v7/src/RField.cxx (hunk 1 above)
Comment (hahnjo) [reply tree/ntuple/v7/src/RField.cxx:3248]: I would advise against this kind of micro-optimizations: Field creation is not on the critical path and we have lots of `virtual` function calls and interpreter invocations here. I also doubt that (integer) modulo operations are still significantly slower than subtraction + bit operation on modern architectures, except for microbenchmarks (maybe). IMHO readability and maintainability should be prioritized.

Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:3248]: I tend to agree here, I'd prioritize maintainability.

**Review by silverweed — APPROVED** [review]

File: tree/ntuple/v7/src/RField.cxx [hunk 3]
```diff
@@ -3208,15 +3225,28 @@ ROOT::Experimental::RVariantField::RVariantField(std::string_view fieldName,
...
+   // With certain template parameters, the union of members of an std::variant starts at an offset > 0.
+   // For instance, std::variant<std::optional<int>> on macOS.
+   auto cl = TClass::GetClass(GetTypeName().c_str());
+   assert(cl);
+   auto dm = reinterpret_cast<TDataMember *>(cl->GetListOfDataMembers()->First());
+   if (dm)
+      fVariantOffset = dm->GetOffset();
+
+   const auto tagSize = GetVariantTagSize();
+   const auto padding = tagSize - (fMaxItemSize % tagSize);
```
Comment (pcanal) [inline tree/ntuple/v7/src/RField.cxx:3248]: Isn't it semantically:
```
const auto padding = tagSize - (fMaxItemSize % alignof<tagtype>);
```
?

Comment (pcanal) [reply tree/ntuple/v7/src/RField.cxx:3248]: If it is and `tagSize` is still 'correct' then a comment is necessary for clarification.

**Pull request merged**